Implement C++ API for defining channel filters.

markdroth opened the pull request:
This compiles, but that's about all I can say for it thus far.  My next step will be to add a test, but I wanted to get your feedback on my overall approach first.

ctiller commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,211 @@
+  static void InitCallElement(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_call_element_args *args) {
+    CallDataType* call_data = elem->call_data;
```
This method should be just:

```c
new (elem->call_data) CallDataType();
```

... there's no guarantee that the data pointed to is initialized, so we want to avoid calling the destructor (or that might crash)

markdroth replied:
Very cool -- I didn't know "new" was able to construct an object in previously-allocated memory.

Done.

ctiller commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,211 @@
+  static void DestroyCallElement(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      const grpc_call_stats *stats, void *and_free_memory) {
+    CallDataType* call_data = elem->call_data;
```
Similarly, this should just be:

reinterpret_cast<CallDataType*>(elem->call_data)->~CallDataType();

markdroth replied:
Duh -- it also didn't occur to me that I could explicitly call the destructor, but it probably should have.

Done.

ctiller commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,211 @@
+  // Destroy() method.
+  virtual ~CallData() {}
+
+  virtual void Destroy() {}
```
No need for Destroy

markdroth replied:
Done.

ctiller commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,211 @@
+  static void InitChannelElement(
+      grpc_exec_ctx *exec_ctx, grpc_channel_element *elem,
+      grpc_channel_element_args *args) {
+    ChannelDataType* channel_data = elem->channel_data;
```
s/CallDataType/ChannelDataType/g for my above comments :)

markdroth replied:
Yup, done.

ctiller commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,211 @@
+// Registers a new filter.
+// Must be called by only one thread at a time.
+template<typename ChannelDataType, typename CallDataType>
+void RegisterChannelFilter(const char* name,
```
We might want to add the argument:
```
std::function<bool(grpc_channel_args*)> should_this_filter_be_included
```
to allow runtime decisions to be made about this filter

markdroth replied:
Done.

ctiller commented:
I think I'm done on the first pass here... looking good.

markdroth commented:
As we discussed, I've changed it to pass the channel args to the ChannelData ctor and the ChannelData object to the CallData ctor.

I've added a test that demonstrates how this works.

Please let me know what you think.  Thanks!

vjpai commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,200 @@
+#include <vector>
+
+#include "src/core/lib/channel/channel_stack.h"
+#include "src/core/lib/surface/channel_init.h"
```
I think "include/grpc++" files should not include "src/..." anything since installations will have the include/ directory but not necessarily the src

markdroth replied:
Yup.  Craig and I talked about this earlier today.  The intent is to move the necessary headers from src to include, thus making them part of the public API.  But Craig wanted to iron out some details of API versioning before we do that.

This PR won't be merged until we have resolved the API versioning issues.

ctiller replied:
Yeah we've been chatting about that offline a little. It forces the issue
of versioning for core. I'll write up a little blurb later on and finally
put a nail in this.

On Thu, Jun 16, 2016, 11:02 AM Vijay Pai <notifications@github.com> wrote:

> In include/grpc++/channel_filter.h
> <https://github.com/grpc/grpc/pull/6915#discussion_r67394362>:
>
> > + * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
> > + * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
> > + * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
> > + *
> > + */
> > +
> > +#ifndef GRPCXX_CHANNEL_FILTER_H
> > +#define GRPCXX_CHANNEL_FILTER_H
> > +
> > +#include <grpc/grpc.h>
> > +
> > +#include <functional>
> > +#include <vector>
> > +
> > +#include "src/core/lib/channel/channel_stack.h"
> > +#include "src/core/lib/surface/channel_init.h"
>
> I think "include/grpc++" files should not include "src/..." anything since
> installations will have the include/ directory but not necessarily the src
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/6915/files/f9c1f7a412619e78a93a116068fe0fe373e172e1#r67394362>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AJpudaecKfZyuwjHvHIt8e-dt6uWcgW8ks5qMY-jgaJpZM4I2y2T>
> .
>

markdroth replied:
As per my discussion with Craig this afternoon, we're not going to make this a public API right now.  So I've moved the header file into the src tree.

vjpai commented on src/cpp/common/channel_filter.cc:
```diff
@@ -0,0 +1,105 @@
+
+namespace internal {
+
+std::vector<FilterRecord> *channel_filters = nullptr;
```
Omit the " = nullptr" - it's more compatible.

markdroth replied:
What's the compatibility issue?

I think this needs to be initialized, since the code in RegisterChannelFilter() depends on the value being nullptr to know whether or not to allocate a new vector object.

vjpai replied:
Some nullptr issues are discussed in https://github.com/grpc/grpc/blob/master/doc/cpp-style-guide.md . I have added this particular one to the big compatibility issue in #6863 . This initialization is not needed because channel_filters here has static lifetime, and all pointers with static lifetime are implicitly initialized to nullptr.

markdroth replied:
Got it.  I think I knew that about variables declared as "static", but I somehow didn't realize that it also applied to those defined in an anonymous namespace.

vjpai replied:
Right - it applies to anything that has static storage, not just the keyword `static`.

vjpai commented on src/cpp/common/channel_filter.cc:
```diff
@@ -0,0 +1,105 @@
+
+bool MaybeAddFilter(grpc_channel_stack_builder *builder, void *arg) {
+  const FilterRecord &filter = *(FilterRecord *)arg;
+  if (filter.include_filter != nullptr) {
```
Omit the "!= nullptr"

markdroth replied:
I tend to find the code more readable this way, because it makes it obvious to the reader than the variable is a pointer (as opposed to, say, a bool).  Is there some compatibility reason to change this?  (For example, if the problem is using nullptr instead of NULL, I could change to the latter.)

vjpai replied:
The problem is discussed in https://github.com/grpc/grpc/blob/master/doc/cpp-style-guide.md :
nullptr wasn't accepted by gcc until 4.6. We can't use NULL as a substitute for nullptr (it's just the numerical value 0)

markdroth replied:
Done.

vjpai commented on test/cpp/end2end/filter_end2end_test.cc:
```diff
@@ -0,0 +1,327 @@
+    // Include a second call to RegisterAsyncGenericService to make sure that
+    // we get an error in the log, since it is not allowed to have 2 async
+    // generic services
+    builder.RegisterAsyncGenericService(&generic_service_);
```
The above seems unneeded, since that's already handled in end2end_test

markdroth replied:
Done.  (This was copied from generic_end2end_test.cc, which is what I used as the base of this file.)

markdroth commented:
This now depends on #7024.

dgquintas commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,365 @@
+#include "src/core/lib/surface/channel_init.h"
+#include "src/core/lib/transport/metadata_batch.h"
+
+//
```
Let's use `///` for comments so that Doxygen picks them up. It's also advisable to use Doxygen tags (https://www.stack.nl/~dimitri/doxygen/manual/commands.html) where appropriate (we use the \ convention).

markdroth replied:
Done.

dgquintas commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,365 @@
+// An interface to define filters.
+//
+// To define a filter, implement a subclass of each of CallData and
+// ChannelData.  Then register the filter using something like this:
```
as a nit/pet-peeve (and with no written rule for nor against it) we've also been use single-space-after-period in docstrings.

markdroth replied:
Sorry, I'm incapable of not typing two spaces after a period. :)  Fixed.

dgquintas commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,365 @@
+ public:
+  explicit MetadataBatch(grpc_metadata_batch *batch) : batch_(batch) {}
+
+  grpc_metadata_batch *batch() const { return batch_; }
```
maybe we'd want a version of `batch()` returning a constant pointer? That'd allow for const instances of `MetadataBatch`.

markdroth replied:
I don't anticipate this being used anywhere other than a filter, and since we don't know whether or not a given filter will want to modify metadata, we need to pass it in as non-const.  So it doesn't seem like there's any immediate benefit in a const method.

dgquintas replied:
Agreed.

dgquintas commented on include/grpc++/channel_filter.h:
```diff
@@ -0,0 +1,365 @@
+  ChannelData(const grpc_channel_args &args, const char *peer) : peer_(peer) {}
+
+ private:
+  const char *peer_;  // Do not own.
```
s/Do/Does? sounds imperative otherwise.

markdroth replied:
Actually, in this particular case, the comment was just plain wrong, so I've removed it and changed the dtor to free the memory. :)

I have updated the wording in several other places, though.

markdroth commented:
@vjpai It looks like the gcc-4.4 compatibility test is failing due to a use of nullptr.  Looking at https://grpc-testing.appspot.com/job/gRPC_portability_pull_requests/4433/language=c++,scenario=linux_x64_gcc4.4_bo/console, the error is:

```
[CXX]     Compiling test/cpp/end2end/generic_end2end_test.cc
test/cpp/end2end/filter_end2end_test.cc: In function 'int main(int, char**)':
test/cpp/end2end/filter_end2end_test.cc:347: error: 'nullptr' was not declared in this scope
make: *** [/var/local/git/grpc/objs/dbg/test/cpp/end2end/filter_end2end_test.o] Error 1
```

In this case, what's happening is that we're got a function that takes a std::function<> as a parameter, and we're calling the function with nullptr as the value for that parameter.  What's the right way to deal with this to make gcc-4.4 happy?  Should I just use NULL instead of nullptr, or is something else needed?

vjpai commented:
I did add a typecast for `nullptr`->`std::function` in PR #6863  so it's definitely feasible. Looking at the error, though, I think I know the issue. Remember that `nullptr` is a defined thing so you have to explicitly `#include <grpc++/support/config.h>` or else `nullptr` won't be understood. Does that work?

vjpai commented:
Also, FWIW, `NULL` won't work. That's just the number 0, and it is not convertible to unique_ptr, function. etc. That was how we got into that hacked-up definition of `nullptr` to begin with.

markdroth commented:
Unfortunately, it looks like adding that include didn't fix the problem.  I'm still seeing exactly the same error:

```
test/cpp/end2end/filter_end2end_test.cc: In function 'int main(int, char**)':
test/cpp/end2end/filter_end2end_test.cc:348: error: 'nullptr' was not declared in this scope
make: *** [/var/local/git/grpc/objs/dbg/test/cpp/end2end/filter_end2end_test.o] Error 1
```

I'm not sure why this isn't working, since I do see the definition of nullptr in config.h.  Any ideas?

markdroth commented:
Oh!  I think the problem is that the definition of nullptr in config.h is in the grpc namespace, and this function call is not.  I'll try moving the call into the grpc namespace and see if that fixes it.

pgrosu commented:
Hi Mark,

GCC 4.4 does not include `nullptr`, and was only introduced in GCC 4.6 as noted in the following link:

https://gcc.gnu.org/gcc-4.6/changes.html

Try using `__null` instead which should work, or define it with something like this:

```C
# define nullptr __null
```

Or more elaborately like this:

```C
#ifndef HAVE_NULLPTR
#if defined(_WIN64)
# define nullptr 0LL
#elif defined(__GNUC__)
# define nullptr __null
#else
# define nullptr 0L
#endif
#endif /* defined(HAVE_NULLPTR) */
```

There are many choices :)

Hope it helps,
Paul

markdroth commented:
@pgrosu, thanks for the info.  I know about the lack of nullptr in gcc-4.4, and @vjpai has created a nullptr implementation that works in gcc-4.4 for cases like this, so the issue here was trying to figure out why that wasn't working.

pgrosu commented:
@markdroth Maybe my eyes are playing tricks on me, but could you please point me to the definition by @vjpai?  I'm looking at [include/grpc++/impl/codegen/config.h](https://github.com/grpc/grpc/blob/master/include/grpc%2B%2B/impl/codegen/config.h#L52-65) at lines 52-65, and not seeing `nullptr`  being specifically defined: 

```
#ifndef __clang__
#ifdef __GNUC__
// nullptr was added in gcc 4.6
#if (__GNUC__ * 100 + __GNUC_MINOR__ < 406)
#define GRPC_CXX0X_NO_NULLPTR 1
#define GRPC_CXX0X_LIMITED_TOSTRING 1
#endif
// final and override were added in gcc 4.7
#if (__GNUC__ * 100 + __GNUC_MINOR__ < 407)
#define GRPC_CXX0X_NO_FINAL 1
#define GRPC_CXX0X_NO_OVERRIDE 1
#endif
#endif
#endif
```

Thanks,
~p

markdroth commented:
@pgrosu, it's here:

https://github.com/grpc/grpc/blob/master/include/grpc%2B%2B/impl/codegen/config.h#L81-109

pgrosu commented:
@markdroth Thank you, but doesn't `#ifdef GRPC_CXX0X_NO_NULLPTR` mean that it will include it if and only if the macro is already defined according to the following:

https://gcc.gnu.org/onlinedocs/cpp/Ifdef.html

I don't believe `GRPC_CXX0X_NO_NULLPTR` is defined anywhere else.

Thanks,
~p

markdroth commented:
@pgrosu, the code you originally pointed to (lines 52-65) cause `GRPC_CXX0X_NO_NULLPTR` to be defined.  Specifically, see line 56.  Thus, when we get to line 81, the definition is present and the `#ifdef` will evaluate to true.

pgrosu commented:
@markdroth Thank you for helping me, up 404 < 406 :)

markdroth commented:
It looks like the portability tests now pass, so moving the use of nullptr into the grpc namespace solved the problem.

pgrosu commented:
Super!

markdroth commented:
retest this please

ctiller commented:
This seems good... I'm not going to get time to do a thorough review (so hoping someone else can pull out the comb), but I have no objections.

markdroth commented:
retest this please

markdroth commented:
David, can you take over this review from Craig, since you've already looked at the code?  Thanks!

markdroth commented:
retest this please

markdroth commented:
retest this please

markdroth commented:
retest this please

dgquintas commented on src/core/lib/channel/context.h:
```diff
@@ -34,10 +34,19 @@
 #define GRPC_CORE_LIB_CHANNEL_CONTEXT_H
 
-/* Call object context pointers */
+// Call object context pointers.
```
Could you use triple slashes so that Doxygen can pick up the comments?

markdroth replied:
Done.

dgquintas commented on src/cpp/common/channel_filter.cc:
```diff
@@ -0,0 +1,120 @@
+
+namespace grpc {
+
+//
```
Not personally against this type of banner comments, but we don't use them across the codebase. If anything, I'd reduce them to a single line with no empty lines in between, viz:

````
// MetadataBatch
grpc_linked_mdelem moar codez ...
```

markdroth replied:
I find these comments useful to help quickly visually find the right section of the file.

Reduced to a single line, but left a blank line afterwards.

dgquintas commented on src/cpp/common/channel_filter.cc:
```diff
@@ -0,0 +1,120 @@
+void ChannelFilterPluginShutdown() {}
+
+}  // namespace internal
+
```
weak nit: remove empty line

markdroth replied:
The two namespaces start at different points, so the blank line in between seems appropriate to me.  (It would be different if, for example, all of our code was in the namespace net::grpc, at which point all files would start and end with the same two lines of namespace wrapping.)

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+#include "src/core/lib/surface/channel_init.h"
+#include "src/core/lib/transport/metadata_batch.h"
+
+///
```
Similar comment as before: we don't usually wrap comments in extra whitespace. You use an extra empty line plus an empty comment line. Perhaps we should standardize on one style on the styleguide :)

markdroth replied:
Removed the extra comment lines.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+
+  grpc_metadata_batch *batch() const { return batch_; }
+
+  // Adds metadata and returns the newly allocated storage.
```
why not `///`?

markdroth replied:
Done.

dgquintas commented on src/cpp/common/channel_filter.cc:
```diff
@@ -0,0 +1,120 @@
+}
+
+//
+// RegisterChannelFilter()
```
there's nothing related to RegisterChannelFilter here.

markdroth replied:
The code here is internal code used by RegisterChannelFilter().  I've updated the comment to make that clear.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+/// A C++ wrapper for the \c grpc_metadata_batch struct.
+class MetadataBatch {
+ public:
+  explicit MetadataBatch(grpc_metadata_batch *batch) : batch_(batch) {}
```
comment plz. Is ownership taken?

markdroth replied:
Done.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+/// A C++ wrapper for the \c grpc_transport_op struct.
+class TransportOp {
+ public:
+  explicit TransportOp(grpc_transport_op *op) : op_(op) {}
```
same. comment wrt ownership of argument.

markdroth replied:
Done.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+ public:
+  explicit TransportStreamOp(grpc_transport_stream_op *op)
+      : op_(op),
+        send_initial_metadata_(op->send_initial_metadata),
```
why do we have all these member variables? they seem to be an alias of their corresponding `op_->` persona.

markdroth replied:
The goal here is to provide a more C++-like API for accessing metadata.  Thus, the member variables are C++ MetadataBatch objects, which are wrapped around the corresponding C grpc_metadata_batch structs.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+
+  MetadataBatch *send_initial_metadata() {
+    return op_->send_initial_metadata == nullptr ? nullptr
+                                                 : &send_initial_metadata_;
```
why not simply return `op_->send_initial_metadata_`?

markdroth replied:
As I mentioned above, the goal here is to provide a more idiomatic C++ API for accessing metadata.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+class CallData {
+ public:
+  virtual ~CallData() {}
+
```
comments on public interfaces please

markdroth replied:
Done.

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+
+namespace internal {
+
+// Defines static members for passing to C core.
```
`s://:///`

markdroth replied:
Since this is just an internal function, not for external use, I was thinking that we don't actually want Doxygen to pick up the docstring for it.  Is there a reason to do so even for internal functions?

dgquintas commented on src/cpp/common/channel_filter.h:
```diff
@@ -0,0 +1,373 @@
+class ChannelFilter GRPC_FINAL {
+ public:
+  static const size_t channel_data_size = sizeof(ChannelDataType);
+
```
comments on publics

markdroth replied:
Added a comment saying that the members of this class correspond to those of the C grpc_channel_filter struct.

dgquintas commented:
LGTM

dgquintas commented:
It seems you need to regenerate projects (https://grpc-testing.appspot.com/job/gRPC_pull_requests/11549/config=opt,language=sanity,platform=linux/consoleText)

markdroth commented:
I believe the remaining test failures are flakes, so I'm calling this ready to merge.

dgquintas commented:
+1
